### PR TITLE
inductor: make as_strided support  non-contiguous input and always fix it's input layout using eager stride

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92247
* __->__ #92063

GIven the following small case:

```
import torch
import torch._dynamo

class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        return torch.as_strided(x + 1, (8, 384, 2, 20, 12), (153600, 1, 61440, 384, 7680))+ 2

x = torch.randn(8, 384, 20, 20).to(memory_format=torch.channels_last)
model= Model().eval()
model = model.to(memory_format=torch.channels_last)
ref = model(x)

with torch.no_grad():
    opt_model = torch._dynamo.optimize('inductor')(model)


with torch.no_grad():
    for i in range(2):
        y1 = opt_model(x)

print(torch.equal(ref, y1))

```

inductor always gets a wrong result:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_xiaobing/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0,
                       float* __restrict__ out_ptr1)
{
    #pragma omp parallel num_threads(40)
    {
        {
            #pragma omp for
            for(long i0=0; i0<8; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<384; i1+=1)
                {
                    #pragma GCC ivdep
                    for(long i2=0; i2<400; i2+=1)
                    {
                        auto tmp0 = in_ptr0[i1 + (384*i2) + (153600*i0)];
                        auto tmp1 = static_cast<float>(1);
                        auto tmp2 = tmp0 + tmp1;
                        out_ptr0[i2 + (400*i1) + (153600*i0)] = tmp2;
                    }
                }
            }
        }
        {
            #pragma omp for  collapse(2)
            for(long i0=0; i0<8; i0+=1)
            {
                for(long i1=0; i1<2; i1+=1)
                {
                    for(long i2=0; i2<5760; i2+=1)
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(out_ptr0 + (16*i2) + (61440*i1) + (153600*i0));
                        auto tmp1 = at::vec::Vectorized<float>(static_cast<float>(2));
                        auto tmp2 = tmp0 + tmp1;
                        tmp2.store(out_ptr1 + (16*i2) + (92160*i1) + (184320*i0));
                    }
                    #pragma omp simd simdlen(8)
                    for(long i2=92160; i2<92160; i2+=1)
                    {
                        auto tmp0 = out_ptr0[i2 + (61440*i1) + (153600*i0)];
                        auto tmp1 = static_cast<float>(2);
                        auto tmp2 = tmp0 + tmp1;
                        out_ptr1[i2 + (92160*i1) + (184320*i0)] = tmp2;
                    }
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    buf0 = empty_strided((8, 384, 20, 20), (153600, 400, 20, 1), device='cpu', dtype=torch.float32)
    buf1 = empty_strided((8, 384, 2, 20, 12), (184320, 1, 92160, 384, 7680), device='cpu', dtype=torch.float32)
    kernel_cpp_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()), c_void_p(buf1.data_ptr()))
    del arg0_1
    return (buf1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((8, 384, 20, 20), (153600, 1, 7680, 384), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))

```

the reason is that there always convert the input to a contiguous layout at **as_strided** lowering step, which is not aligned with the eager model input stride.

```
class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[8, 384, 20, 20]):
        # File: model_test.py:52, code: return torch.as_strided(x + 1, (8, 384, 2, 20, 12), (153600, 1, 61440, 384, 7680))+ 2
        add: f32[8, 384, 20, 20] = torch.ops.aten.add.Tensor(arg0_1, 1);  arg0_1 = None
        as_strided: f32[8, 384, 2, 20, 12] = torch.ops.aten.as_strided.default(add, [8, 384, 2, 20, 12], [153600, 1, 61440, 384, 7680]);  add = None
        add_1: f32[8, 384, 2, 20, 12] = torch.ops.aten.add.Tensor(as_strided, 2);  as_strided = None
        return (add_1,)

```

This PR will always fix **as_strided** stride with eager model's stride, and also make **as_strided** support channels_last input:

```
from ctypes import c_void_p, c_long
import torch
import random
from torch import empty_strided, as_strided, device
from torch._inductor.codecache import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels

aten = torch.ops.aten
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
async_compile = AsyncCompile()


kernel_cpp_0 = async_compile.cpp('''
#include "/tmp/torchinductor_xiaobing/77/c7773nj5pwikpmm2pwa62rcudlf7p3if7eyqb5k4sjsvewwje4le.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0,
                       float* __restrict__ out_ptr1)
{
    #pragma omp parallel num_threads(40)
    {
        {
            #pragma omp for
            for(long i0=0; i0<76800; i0+=1)
            {
                auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + 16*i0);
                auto tmp1 = at::vec::Vectorized<float>(static_cast<float>(1));
                auto tmp2 = tmp0 + tmp1;
                tmp2.store(out_ptr0 + 16*i0);
            }
            #pragma omp for simd simdlen(8)
            for(long i0=1228800; i0<1228800; i0+=1)
            {
                auto tmp0 = in_ptr0[i0];
                auto tmp1 = static_cast<float>(1);
                auto tmp2 = tmp0 + tmp1;
                out_ptr0[i0] = tmp2;
            }
        }
        {
            #pragma omp for  collapse(2)
            for(long i0=0; i0<8; i0+=1)
            {
                for(long i1=0; i1<2; i1+=1)
                {
                    for(long i2=0; i2<5760; i2+=1)
                    {
                        auto tmp0 = at::vec::Vectorized<float>::loadu(out_ptr0 + (16*i2) + (61440*i1) + (153600*i0));
                        auto tmp1 = at::vec::Vectorized<float>(static_cast<float>(2));
                        auto tmp2 = tmp0 + tmp1;
                        tmp2.store(out_ptr1 + (16*i2) + (92160*i1) + (184320*i0));
                    }
                    #pragma omp simd simdlen(8)
                    for(long i2=92160; i2<92160; i2+=1)
                    {
                        auto tmp0 = out_ptr0[i2 + (61440*i1) + (153600*i0)];
                        auto tmp1 = static_cast<float>(2);
                        auto tmp2 = tmp0 + tmp1;
                        out_ptr1[i2 + (92160*i1) + (184320*i0)] = tmp2;
                    }
                }
            }
        }
    }
}
''')


async_compile.wait(globals())
del async_compile

def call(args):
    arg0_1, = args
    args.clear()
    buf0 = empty_strided((8, 384, 20, 20), (153600, 1, 7680, 384), device='cpu', dtype=torch.float32)
    buf1 = empty_strided((8, 384, 2, 20, 12), (184320, 1, 92160, 384, 7680), device='cpu', dtype=torch.float32)
    kernel_cpp_0(c_void_p(arg0_1.data_ptr()), c_void_p(buf0.data_ptr()), c_void_p(buf1.data_ptr()))
    del arg0_1
    return (buf1, )


if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    arg0_1 = rand_strided((8, 384, 20, 20), (153600, 1, 7680, 384), device='cpu', dtype=torch.float32)
    print_performance(lambda: call([arg0_1]))
```

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
This PR will also

File: torch/_inductor/graph.py [hunk 1]
```diff
@@ -385,7 +385,7 @@ def run_node(self, n: torch.fx.Node):
             # Realize if (1) any user need inputs realized, or (2) there is
             # already too many reads and rematerializing can be bad.
             num_users = len(set(n.users))
-            if num_users > 1 and isinstance(result, TensorBox):
+            if num_users > 0 and isinstance(result, TensorBox):
```
Comment (XiaobingSuper) [inline torch/_inductor/graph.py:388]: Please fix me, I think we should do the realization even with only one user?

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -734,7 +734,7 @@ def as_strided(x, size, stride, storage_offset=None):
         # as_strided ignores views
         x = x.data.unwrap_view()
     x.realize()
-    if not ir.is_contiguous_storage_and_layout(x):
+    if not ir.is_contiguous_storage_and_layout(x, check_channels_last=True):
         raise NotImplementedError(f"unrealized as_strided({x}, ...)")
     storage, old_layout = ir.as_contiguous_storage_and_layout(x)
```
Comment (jgong5) [inline torch/_inductor/lowering.py:739]: How about we explicitly check if `x` has fixed layout, and get its layout directly here?

**Comment by XiaobingSuper** [comment]
@jansel , @desertfire, @Chillee, please help review this PR, this PR can help fix Timm eca_halonext26ts accuracy issue when calling channels last path. Thanks!

**Review by desertfire — COMMENTED** [review]
@eellison , do you think this case should be covered by your graph output layout enforcement?

File: torch/_inductor/graph.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/graph.py:388]: I didn't test the performance on GPU side, how about separate **as_strided**'s input realization from this path? for **as_strided**, its input should be always realized "eagerly".

Comment (desertfire) [reply torch/_inductor/graph.py:388]: I am ok with that. Please make sure to add comments to explain it.

Comment (eellison) [inline torch/_inductor/graph.py:388]: I'm a little wary of changing recomputation heuristics in this PR. Could you change in a separate pr please ?

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -1007,10 +1007,13 @@ def is_storage_and_layout(x):
         return False
 
 
-def is_contiguous_storage_and_layout(x):
+def is_contiguous_storage_and_layout(x, check_channels_last=False):
     try:
         buffer, layout = as_storage_and_layout(x, freeze=False)
-        return layout.is_contiguous()
+        is_contiguous = layout.is_contiguous()
+        if check_channels_last:
```
Comment (eellison) [inline torch/_inductor/ir.py:1014]: nit: 
`return layout.is_contiguous() or (check_channels_last and layout.is_channels_last_contiguous())`

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/lowering.py:740]: I think you can call `as_storage_and_layout` to avoid forcing contiguity

**Review by eellison — COMMENTED** [review]
Looks good to me, but needs the recomputation fix you and bin were discussing. Some of the as_strided issues might be fixed by https://github.com/pytorch/pytorch/pull/92255.

@desertfire I don't think this would be handled by the graph output stride constraints, since that is more about a reshape correctness that occurs outside of inductor, where as this is a semantic mismatch that occurs internally

**Review by jansel — CHANGES_REQUESTED** [review]
I think there should be a more general fix to this that doesn't need to check channels last specifically.

I think we are handling as_strided wrong.  I'd suggest changing to to apply a FixedLayout to the input based on the strides from eager in `node.meta["val"]`.

File: torch/_inductor/graph.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/graph.py]: Separate as_strided's input realization from this path and adds some comments.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/graph.py [hunk 4]
```diff
@@ -368,11 +368,16 @@ def run_node(self, n: torch.fx.Node):
...
-            # so that user-land view() will not throw because inductor
+            # 1. user-land view() will not throw because inductor
             # output different strides than eager
             # long term the solution is to make view() always succeed
             # with infallible strides.
-            if any(user.op == "output" for user in n.users):
+            # 2: as_strided op, we need make sure its input has same size/stride with
+            # eager model to align with eager behavior.
+            if any(
+                user.op == "output" or user.target == torch.ops.aten.as_strided.default
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/graph.py:378]: Yes, those two ops are added.
